### DISCUSSION TITLE
Redirect to /login page if not logged in?

### DISCUSSION THREAD

**Discussion by fmaritato** [apache/superset#14938] [Q&A / Help]
**Is your feature request related to a problem? Please describe.**
We use okta for authentication, but right now, you have to know to click the login link in the upper right before you can access any dashboards. It's super confusing to our non-technical people who just get a link to a dashboard because it just shows a bunch of errors if they arent logged in.

**Describe the solution you'd like**
Is there a way to force superset to redirect to the /login page if they are not logged in?
I am using a custom superset_config.py and custom_sso_security_manager.py for okta anyway, maybe there is some way to accomplish this in there?

**Comment by nytai** [comment]
You should be able to achieve this with the custom security manager. I'm not sure how you're loading the user, but it should be simple to add a redirect in that method if no user is found.

**Reply by fmaritato** [reply]
@nytai Are there examples around on how to do this? 

I'm overriding auth_user_oauth and oauth_user_info methods in my security manager.  If I'm logged out and just navigate to the dashboards page these functions arent called. I see they are called only when the login action is done through the ui.

```
class CustomSsoSecurityManager(SupersetSecurityManager):
    def __init__(self, appbuilder):
        super().__init__(appbuilder)
        self.OKTA_BASE_URL = os.environ.get('OKTA_BASE_URL')
        self.userinfo_uri = f"{self.OKTA_BASE_URL}userinfo"

    def oauth_user_info(self, provider, response=None):
        if provider == 'okta':
            response = self.appbuilder.sm.oauth_remotes[provider].get(self.userinfo_uri)
            res = response.json()
            logger.error(f"user_data: {json.dumps(res)}")
            groups = []
            if 'groups' in res:
                for group in res['groups']:
                    if group == 'APP-SUPERSET-ADMIN':
                        groups.append("Admin")
            if len(groups) == 0:
                groups.append("Public")

            return_val = {
                'username': res['preferred_username'],
                'name': res['name'],
                'email': res['email'],
                'first_name': res['given_name'],
                'last_name': res['family_name'],
                'roles': groups,
            }
            logger.debug(f"ReturnVal: {return_val}")
            return return_val

    def auth_user_oauth(self, userinfo):
        user = super(CustomSsoSecurityManager, self).auth_user_oauth(userinfo)
        logger.error(f"user: {user}")
        roles = [self.find_role(x) for x in userinfo['roles']]
        roles = [x for x in roles if x is not None]
        user.roles = roles
        logger.debug(' Update <User: %s> role to %s', user.username, roles)
        self.update_user(user)  # update user roles
        return user

```

**Comment by toransahu** [comment]
@fmaritato Did you find a solution to this? I'm also facing the same issue.

**Comment by strangerDevloper** [comment]
@toransahu , You can refer #30380 this PR

**Reply by toransahu** [reply]
Thanks @strangerDevloper
